File: tasks/implementations/node/tasks/decision_tree.js
const fs = require('fs');
const THRESHOLDS = [250, 500, 750];
const MAX_DEPTH = 3;

function gini(rows) {
  if (rows.length === 0) return 0;
  let ones = 0;
  for (const row of rows) ones += row[3];
  const zeros = rows.length - ones;
  const p0 = zeros / rows.length;
  const p1 = ones / rows.length;
  return 1 - p0 * p0 - p1 * p1;
}

function majority(rows) {
  let ones = 0;
  for (const row of rows) ones += row[3];
  return ones * 2 >= rows.length ? 1 : 0;
}

function build(rows, depth) {
  if (depth === 0 || rows.length === 0) return { kind: 'leaf', value: rows.length ? majority(rows) : 0 };
  if (rows.every((row) => row[3] === rows[0][3])) return { kind: 'leaf', value: rows[0][3] };
  let best = null;
  let bestScore = Infinity;
  for (let feature = 0; feature < 3; feature++) {
    for (const threshold of THRESHOLDS) {
      const left = [];
      const right = [];
      for (const row of rows) {
        if (row[feature] <= threshold) left.push(row); else right.push(row);
      }
      if (!left.length || !right.length) continue;
      const score = (left.length * gini(left) + right.length * gini(right)) / rows.length;
      if (score < bestScore) {
        bestScore = score;
        best = { feature, threshold, left, right };
      }
    }
  }
  if (!best) return { kind: 'leaf', value: majority(rows) };
  return { kind: 'node', feature: best.feature, threshold: best.threshold, left: build(best.left, depth - 1), right: build(best.right, depth - 1) };
}

function predict(tree, row) {
  if (tree.kind === 'leaf') return tree.value;
  return predict(row[tree.feature] <= tree.threshold ? tree.left : tree.right, row);
}

module.exports = function(size, fixturesRoot) {
  const lines = fs.readFileSync(`${fixturesRoot}/generated/decision_tree/${size}.csv`, 'utf8').trim().split(/\r?\n/);
  const rows = [];
  for (let i = 1; i < lines.length; i++) rows.push(lines[i].split(',').map(Number));
  const tree = build(rows, MAX_DEPTH);
  let correct = 0;
  let predSum = 0;
  for (let i = 0; i < rows.length; i++) {
    const pred = predict(tree, rows[i]);
    if (pred === rows[i][3]) correct += 1;
    predSum += pred * (i + 1);
  }
  console.log(correct * 100000 + predSum);
};

File: tasks/implementations/rust/tasks/decision_tree.rs
use std::fs;

const THRESHOLDS: [i64; 3] = [250, 500, 750];

#[derive(Clone)]
enum Node {
    Leaf(i64),
    Split { feature: usize, threshold: i64, left: Box<Node>, right: Box<Node> },
}

fn gini(rows: &[Vec<i64>]) -> f64 {
    if rows.is_empty() { return 0.0; }
    let ones: i64 = rows.iter().map(|row| row[3]).sum();
    let zeros = rows.len() as i64 - ones;
    let p0 = zeros as f64 / rows.len() as f64;
    let p1 = ones as f64 / rows.len() as f64;
    1.0 - p0 * p0 - p1 * p1
}

fn majority(rows: &[Vec<i64>]) -> i64 {
    let ones: i64 = rows.iter().map(|row| row[3]).sum();
    if ones * 2 >= rows.len() as i64 { 1 } else { 0 }
}

fn build(rows: &[Vec<i64>], depth: usize) -> Node {
    if depth == 0 || rows.is_empty() { return Node::Leaf(if rows.is_empty() { 0 } else { majority(rows) }); }
    if rows.iter().all(|row| row[3] == rows[0][3]) { return Node::Leaf(rows[0][3]); }
    let mut best: Option<(usize, i64, Vec<Vec<i64>>, Vec<Vec<i64>>)> = None;
    let mut best_score = f64::INFINITY;
    for feature in 0..3 {
        for threshold in THRESHOLDS {
            let mut left = Vec::new();
            let mut right = Vec::new();
            for row in rows {
                if row[feature] <= threshold { left.push(row.clone()); } else { right.push(row.clone()); }
            }
            if left.is_empty() || right.is_empty() { continue; }
            let score = (left.len() as f64 * gini(&left) + right.len() as f64 * gini(&right)) / rows.len() as f64;
            if score < best_score {
                best_score = score;
                best = Some((feature, threshold, left, right));
            }
        }
    }
    if let Some((feature, threshold, left, right)) = best {
        Node::Split { feature, threshold, left: Box::new(build(&left, depth - 1)), right: Box::new(build(&right, depth - 1)) }
    } else {
        Node::Leaf(majority(rows))
    }
}

fn predict(node: &Node, row: &[i64]) -> i64 {
    match node {
        Node::Leaf(value) => *value,
        Node::Split { feature, threshold, left, right } => {
            if row[*feature] <= *threshold { predict(left, row) } else { predict(right, row) }
        }
    }
}

pub fn run(size: &str, fixtures_root: &str) {
    let text = fs::read_to_string(format!("{}/generated/decision_tree/{}.csv", fixtures_root, size)).unwrap();
    let mut rows: Vec<Vec<i64>> = Vec::new();
    for line in text.lines().skip(1) {
        rows.push(line.split(',').map(|x| x.parse::<i64>().unwrap()).collect());
    }
    let tree = build(&rows, 3);
    let mut correct = 0i64;
    let mut pred_sum = 0i64;
    for (idx, row) in rows.iter().enumerate() {
        let pred = predict(&tree, row);
        if pred == row[3] { correct += 1; }
        pred_sum += pred * (idx as i64 + 1);
    }
    println!("{}", correct * 100000 + pred_sum);
}
